### ISSUE TITLE
Slow restarts

### ISSUE THREAD

**Issue by nearapogee** [puma/puma#1730]
We love puma and have been using it for years... baffled by this one though! What are we missing?

### Steps to reproduce

1) puma configs
```ruby
# web
tag "#{ENV['RAILS_ENV']}-app-puma"
environment ENV['RAILS_ENV']

pidfile "#{ENV['SHARED_PATH']}/puma.pid"
stdout_redirect "#{ENV['SHARED_PATH']}/log/#{ENV['RAILS_ENV']}_app_puma.stdout.log",
  "#{ENV['SHARED_PATH']}/log/#{ENV['RAILS_ENV']}_app_puma.stderr.log",
  false # don't append

bind "unix://#{ENV['SHARED_PATH']}/#{ENV['RAILS_ENV']}-app-puma.sock"
worker_timeout 30

threads 0,4
workers 1

preload_app!
on_worker_boot do
  ActiveSupport.on_load(:active_record) do
    ActiveRecord::Base.establish_connection
  end
  Que.mode = :off
end
```

``` ruby
# action cable
tag "#{ENV['RAILS_ENV']}-app-puma-actioncable"
environment ENV['RAILS_ENV']

pidfile "#{ENV['SHARED_PATH']}/puma-actioncable.pid"
stdout_redirect "#{ENV['SHARED_PATH']}/log/#{ENV['RAILS_ENV']}_app_puma_actioncable.stdout.log",
  "#{ENV['SHARED_PATH']}/log/#{ENV['RAILS_ENV']}_app_puma_actioncable.stderr.log",
  false # don't append

bind "unix://#{ENV['SHARED_PATH']}/#{ENV['RAILS_ENV']}-app-puma-actioncable.sock"
worker_timeout 30

threads 1,4
workers 1
```

2) Action Cable Configs

All defaults except for:
``` ruby
  config.action_cable.allowed_request_origins = ['https://www.app.com']
  config.action_cable.logger = ActiveSupport::Logger.new(
    File.join(Rails.root, 'log', "#{Rails.env}-cable.log")
  )

```

3) Systemd
```
# web
[Unit]
Description=Rails Web Server
PartOf=staging-app.target

[Service]
EnvironmentFile=/etc/default/staging-app

User=staging-app
WorkingDirectory=/home/staging-app/app/current
Type=simple
Restart=always
KillMode=process

ExecStart=/bin/bash -lc 'bundle exec --keep-file-descriptors puma -C config/deploy/staging/puma.rb'
```

```
# cable
[Unit]
Description=Rails Cable Server
PartOf=staging-app.target

[Service]
EnvironmentFile=/etc/default/staging-app

User=staging-app
WorkingDirectory=/home/staging-app/app/current
Type=simple
Restart=always
KillMode=process

ExecStart=/bin/bash -lc 'bundle exec --keep-file-descriptors puma -C config/deploy/staging/puma-action-cable.rb'
```

4) Restart
```
time sudo systemctl restart staging-app-puma.service

real	1m30.106s
user	0m0.011s
sys	0m0.000s
```

### Expected behavior

Normally, we get instant restarts with puma.

### Actual behavior

Restarts take 1m30s usually.

I have never seen it take this long, but we weren't the original developer of the application.

This application uses ActionCable running in a separate process with nginx sending all /cable requests to another process (also on puma). Both processes take the 1m30s to restart, the ActionCable config file is the same minus `preload_app!` and the `on_worker_boot` block.

The ActiveJob process restarts instantly.

#### Questions
1) Is there a way to tell what is happening in puma after the restart is initiated?
2) Is it expected Action Cable processes take a long time to stop restart? But with no Action Cable requests going to the puma web process, why is that slow to restart as well?
3) Is this a known problem with 5.1? The client will be updating eventually.
4) What are we missing?

### System configuration

**Ruby version**: 2.6.0
**Rails version**: 5.1.6.1
**Puma version**: 3.12.0

**Comment by themilkman** [comment]
We noticed similar issues on a few of our servers (rails >=5.2, ruby >= 2.6.0).

**Comment by rromanchuk** [comment]
@themilkman @nearapogee noticing similar, also using ActionCable, shared. 

Rails 5.2.2
Ruby 2.6.1
puma 3.12.0

Let me gather my config (i'm using pretty vanilla systemd + socket) info and do some data collection and i'll report back

**Comment by themilkman** [comment]
We don't use ActionCable in our apps.

But also a trivial systemd+socket setup.

**Comment by rromanchuk** [comment]
I just switched to the alternative forked `systemd` in order to utilize phased-restarts  with `pumactl` https://github.com/puma/puma/blob/master/docs/systemd.md#alternative-forking-configuration It was a huge pain, and somewhat fragile compared to the systemd simple/socket I'm going to do a quick sanity test with phased restarts and see how it handles. 

hot restarts were dragggggging causing racktimeouts and gateway timeouts from the loadbalancer with two workers 16 threads, `preload_app`.  My guess is with preload, and perhaps bad actors preventing connections from being drained, workers cant reboot before timeout threshholds are hit. In fact i wouldn't be surprised if my timeouts were the only thing allowing the restart to complete

I'll post all me new/old configs in a bit, it's a painstaking process to debug this in production env.

**Comment by calmyournerves** [comment]
Not sure if related, but we have seen our restart times go up significantly (from a few seconds to >30s) after upgrading to Ruby 2.6(.1) from 2.5.

**Comment by Zyndoras** [comment]
Restart times went up to (exactly) 90 seconds for each app that i updated from ruby 2.5 to 2.6.
The issue persists when updating to 2.6.1.

**Comment by MSP-Greg** [comment]
JFYI, Using Ruby 2.6 (and earlier, trunk) Puma Travis CI has been failing on tests related to 'clustered restart' since last year.

Since the tests require `fork`, the tests in question are skipped on Appveyor/Windows.

**Comment by evanphx** [comment]
This is almost certainly the ruby 2.6.0 wait bug. Why are you doing `workers 1`? Might as well avoid the overhead of using the clustering in that case and just comment it out and use the single process model.

**Comment by nammosov** [comment]
I have the same issue with ruby 2.6.2
90 seconds is a default timeout for systemd restart command, after that process is just killed by the systemd and started again

**Comment by MSP-Greg** [comment]
I believe master has a workaround for the 2.6 bug, 3.12.1 does not.  If possible, results with master might be helpful...

**Comment by dekellum** [comment]
@evanphx Is the ruby 2.6.0 wait bug https://bugs.ruby-lang.org/issues/14867 or something else?

**Comment by MSP-Greg** [comment]
It's mentioned somewhere here.  I think 14867 is only an issue with JIT enabled

**Comment by MSP-Greg** [comment]
In my fork, I just bypassed the 2.6 workaround on top of PR #1760, and both 2.6.3 & trunk failed in Travis.  Hence, for the time being, the workaround in master is still required...

**Comment by dentarg** [comment]
I'm not sure there is an open bug for the "wait bug", but I commented on https://bugs.ruby-lang.org/issues/15499 when I saw problems with Puma (cluster mode) and Ruby 2.6.0.

I have [some code](https://github.com/dentarg/gists/tree/master/gists/ruby-bug-15499) that reproduces the issue I was seeing (checked ruby-trunk a few weeks back). I've been thinking that I should try to reproduce it without depending on Puma, but I haven't had much time to try that. Maybe someone should just file a new bug with the info we got already?
